### PR TITLE
JIT: add suppressed zero inits if we transform tail recursive call to…

### PR DESCRIPTION
**Pull request by AndyAyersMS**
… jump

Track whether the jit has suppressed in-body zero initialization for any locals
because at the time it appeared prolog initialization would be sufficient. If
the jit subsequently decides to change a tail recursive call into a loop, use
this information to add back the missing zero initializations.

Added a test case.

Addresses #33529.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

No diffs on x64 FX -- not too surprising as we think this case can't be reached by C#.

File: src/coreclr/src/jit/morph.cpp [hunk 1]
```diff
@@ -8159,7 +8159,8 @@ void Compiler::fgMorphRecursiveFastTailCallIntoLoop(BasicBlock* block, GenTreeCa
                 var_types lclType            = varDsc->TypeGet();
                 bool      isUserLocal        = (varNum < info.compLocalsCount);
                 bool      structWithGCFields = ((lclType == TYP_STRUCT) && varDsc->GetLayout()->HasGCPtr());
-                if (isUserLocal || structWithGCFields)
+                bool      hadSuppressedInit  = varDsc->lvSuppressedZeroInit;
+                if (isUserLocal || structWithGCFields || hadSuppressedInit)
```
Comment (erozenfeld) [inline src/coreclr/src/jit/morph.cpp:8163]: I think this is too conservative. Consider this case:
`info.compInitMem` is false
`compSuppressedZeroInit` is true
`hadSuppressedInit` is false

In this case we shouldn't add zero init regardless of the value of `isUserLocal` and `structWithGCFields`.

Comment (AndyAyersMS) [reply src/coreclr/src/jit/morph.cpp:8163]: Sure, will update.

**Review by erozenfeld — APPROVED** [review]
LGTM
Thank you for fixing this!

**Comment by AndyAyersMS** [comment]
Can't figure out how to rerun just the failed parts, so am going to bounce this to restart CI.

**Pull request merged**